Create initial page for `CS8422` (#38813)
dotnet/docs: docs/csharp/language-reference/compiler-messages/cs8422.md

---
description: "Compiler Error CS8422"
title: "Compiler Error CS8422"
ms.date: 12/17/2023
f1_keywords:
  - "CS8422"
helpviewer_keywords:
  - "CS8422"
---
# Compiler Error CS8422

A static local function cannot contain a reference to 'this' or 'base'.

> As of C# 8.0 [`static`](../keywords/static.md) keyword is allowed to be used with local functions to prevent from accidentially shadowing or anyhow modifying the state of parent methods or instances fields.

Reviewer: a couple wording nits:

```suggestion
The [`static`](../keywords/static.md) keyword on a local function prevents the local function from accessing the state of parent instance methods or instances fields.
```

> <br/>This error is to indicate about such scenario, of mixing contexts, taking place.

Reviewer: Change to present tense: and active voice:

```suggestion
<br/>This error indicates the local function accesses instance members of the containing type.
```

## Example

The following sample generates CS8422:

```csharp
public class C
{
    private int counter = 1;
    public void IncreaseCounter()
    {
        static void localFunc(int addition)
        {
            this.counter += addition;   // CS8422 due to reference to 'this'
            base.ToString();            // CS8422 due to reference to 'base'

            // Also for implicit 'this' or 'base' references:
            counter += addition;   // CS8422 due to implicit reference to 'this'
            ToString();            // CS8422 due to implicit reference to 'base'
        }
        localFunc(10);
        Console.WriteLine(this.counter);
    }
}
```

## To correct this error

> If there is an intention of local function to capture and modify the state of a parent then it shouldn't be declared static:

Reviewer: I'd also add a sentence that you could remove the access to instance members:

```suggestion
If there is an intention of local function to capture and modify the state of a parent then it shouldn't be declared static. You could also remove access to any instance members.
```
